mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,346 @@
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct SplitMultipleTapesPass : public impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass> {
+    using impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass>::SplitMultipleTapesPassBase;
+
+    unsigned int countTapes(const func::FuncOp &func)
+    {
+        // Count the number of quantum.device operations in a function
+        unsigned int count = 0;
+        func->walk([&](catalyst::quantum::DeviceInitOp op) { count++; });
+        return count;
+    } // countTapes()
+
+    void collectOperationsForEachTape(const func::FuncOp &func,
+                                      SmallVector<std::vector<Operation *>> &OpsEachTape)
+    {
+        // During tracing, each tape starts with a qdevice_p primitive
+        // This means each tape starts with a quantum.device
+        // and ends with a quantum.device_release
+
+        // The structure of the FuncOp looks like the following:
+        // func.func @circuit(...) -> ... {...} {
+        // preprocessing
+        // quantum.device[...]
+        // tape 1
+        // quantum.device_release
+        // quantum.device[...]
+        // tape 2
+        // quantum.device_release
+        // ... more tapes
+        // post processing and return
+        // }
+
+        // This function parses the operations in the funcop and puts them into
+        // the container as
+        // OpsEachTape = [[preprocessing ops], [tape1 ops], [tape2 ops], ..., [postprocessing ops]]
+
+        // special: [0] for pre and [-1] for post processing
+        unsigned int curTape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (isa<catalyst::quantum::DeviceInitOp>(op)) {
+                curTape++;
+            }
+            OpsEachTape[curTape].push_back(op);
+            if ((isa<catalyst::quantum::DeviceReleaseOp>(op)) &&
+                (curTape == OpsEachTape.size() - 2)) {
+                // reached post processing
+                curTape++;
+            }
+        });
+    } // collectOperationsForEachTape()
+
+    unsigned int getResultNumber(Operation *op, Value v)
+    {
+        // Get the result index i such that v = op->getResult(i)
+        for (unsigned int i = 0; i < op->getNumResults(); i++) {
+            if (op->getResult(i) == v) {
+                return i;
+            }
+        }
+        assert(false && "Value is not a result of the operation!");
+    } // getResultNumber()
+
+    void collectNecessaryValuesFromEarlierTapes(
+        const std::vector<Operation *> &TapeOps,
+        SmallSet<std::pair<Operation *, unsigned int>, SmallSetSize>
+            &NecessaryValuesFromEarlierTapes)
+    {
+        // Go through a list of operations and collect all the necessary operand values
+        // not defined in this list itself, and add them to NecessaryValuesFromEarlierTapes
+        for (Operation *PPOp : TapeOps) {
```

Comment:
Is this paragraph a bit "too-nested"? It's a bit hard to read for my eyes : )

I can factor out the lambda, early `continue` instead of nested `if`s, etc., to make it a bit prettier

Reply:
done